savefile/savefile.go

```diff
@@ -0,0 +1,129 @@
+package savefile
+
+import (
+	"bytes"
+	"compress/zlib"
+	"encoding/binary"
+	"errors"
+	"io"
+	"io/ioutil"
+
+	"github.com/sector-f/jhmod/nvc"
+)
+
+const (
+	// Magic header for Jupiter Hell save save files.
+	magic = "\xde\xc0\xad\xde"
+)
+
+type savedata struct {
+	// Player name
+	PlayerName string
+	// Game mode.  This can be "jh", and various others.
+	GameMode string
+	// The current level's name.
+	CurrentLevel string
+	// The seed used to generate the game.
+	Seed uint32
+}
+
+// Something was reading for data but got an EOF.
+var UnexpectedEndOfFileErr error = errors.New("unexpected eof")
+
+// General parse error.  TODO make more descriptive errors.
+var SaveParseErr error = errors.New("general parse error")
+
+// Read in a string from a buffer of size suffixed with a length byte.
+func readString(r io.Reader, size int) (string, error) {
+	buf := make([]byte, size)
+	nRead, err := r.Read(buf[:])
+	if err != nil {
+		return "", err
+	}
+	if nRead != size {
+		return "", UnexpectedEndOfFileErr
+	}
+	recordedSize := buf[size-1]
+	if int(recordedSize) > size {
+		return "", SaveParseErr
+	}
+	stringBytes := buf[:recordedSize]
+
+	for _, b := range stringBytes {
+		if b == 0 {
+			return "", SaveParseErr
+		}
+	}
+	// TODO verify valid text.
+
+	return string(stringBytes), nil
+}
+
+// Advance the reader r by size bytes.
+func discard(r io.Reader, size int) error {
+	buf := make([]byte, size)
+	_, err := io.ReadFull(r, buf[:])
```

Comment:
Nit: this could be done with a for loop to avoid allocating memory that's actually getting used for anything

It's also worth pointing out that `bufio.Reader` has a `Discard` method to achieve this (although using a `bufio.Reader` here might be overkill?)

Reply:
i forgot why i avoided bufio but there was a reason!  We could use bufio if you think we need it

Reply:
It looks like save files are <100 KB, so there's honestly no real need to replace this function with code that uses less memory :man_shrugging: